### PR TITLE
Mute users who spam/don't attend for some time (#15)

### PR DESCRIPTION
**Pull request by Chabare**
Spam rules discussed in #15

### CODE REVIEW

**Comment by BjoernPetersen** [comment]
Why is this PR labeled "good first issue"? I don't even think the corresponding issue should have that label, as it's kind of an important design decision.

File: dicers_bot/bot.py [hunk 1]
```diff
@@ -105,18 +105,21 @@ def register_main(self, update: Update):
...
         self.save_state()
 
+    def mute_user(self, chat_id: str, user_id: str, until_date):
+        try:
+            self.updater.bot.restrict_chat_member(chat_id, user_id, can_send_messages=False,
+                                                  until_date=until_date)
+        except Exception as e:
+            # Admins can't be banned
+            return False
+        self.updater.bot.send_message(chat_id=chat_id, text="User {} has been banned for 2 hours.".format(
```
Comment (BjoernPetersen) [inline dicers_bot/bot.py]:
```suggestion
        self.updater.bot.send_message(chat_id=chat_id, text="User {} has been muted for 2 hours.".format(
```

File: dicers_bot/bot.py [hunk 2]
```diff
@@ -105,18 +105,21 @@ def register_main(self, update: Update):
...
+                self.mute_user(chat_id, user_id, until_date=until_date)
 
         self.save_state()
 
+    def mute_user(self, chat_id: str, user_id: str, until_date):
+        try:
+            self.updater.bot.restrict_chat_member(chat_id, user_id, can_send_messages=False,
+                                                  until_date=until_date)
+        except Exception as e:
+            # Admins can't be banned
```
Comment (BjoernPetersen) [inline dicers_bot/bot.py]:
```suggestion
            # Admins can't be muted
```

**Comment by Chabare** [comment]
The scope of this branch kinda got away from me :sweat_smile:

**Comment by BjoernPetersen** [comment]
Setting the correct timezone sounds like heresy to me

**Comment by BjoernPetersen** [comment]
BTW, shouldn't that WIP app block this PR right now?

**Comment by Chabare** [comment]
:joy::joy::joy: Yeah, I needed to convince myself too..

---

Theoretically it should block this PR. I haven't found anything which would block the integration from working. Maybe it's a sign ;)

File: main.py [hunk 3]
```diff
@@ -1,22 +1,42 @@
...
+    schedule.every().monday.at(attend_time)
+    schedule.every().monday.at(dice_time)
+    schedule.every().tuesday.at(reset_time)
+
+    logger.info("Run schedule")
     while True:
         schedule.run_pending()
         time.sleep(5)
 
+    logger.error("run_pending for schedule has ended")
```
Comment (BjoernPetersen) [inline main.py]: Unreachable

File: main.py [hunk 4]
```diff
@@ -1,22 +1,42 @@
...
+    logger.debug("Start run_scheduler")
+    default_timezone = timezone("Europe/Berlin")
+    today: datetime = datetime.today()
+    attend_time: str = default_timezone.localize(today.replace(hour=14, minute=0)).strftime("%H:%M")
+    reset_time: str = default_timezone.localize(today.replace(hour=0, minute=0)).strftime("%H:%M")
+    dice_time: str = default_timezone.localize(today.replace(hour=20, minute=30)).strftime("%H:%M")
+    logger.info("Set schedule")
+    schedule.every().monday.at(attend_time)
+    schedule.every().monday.at(dice_time)
+    schedule.every().tuesday.at(reset_time)
```
Comment (BjoernPetersen) [inline main.py]: These 3 lines are missing a `.do(<job>)`

File: dicers_bot/bot.py [hunk 5]
```diff
@@ -97,26 +105,61 @@ def register_main(self, update: Update):
...
+                message = "Sadly, user {} couldn't be restricted due to: `{}`. Shame on {}".format(user.name,
+                                                                                                   e.message,
+                                                                                                   user.name)
+                self.logger.info("{}".format(message))
+                self.updater.bot.send_message(chat_id=chat_id, text=message, parse_mode=ParseMode.MARKDOWN)
+            self.logger.error(e)
+            result = False
+
+        try:
+            self.chats.get(chat_id).get_user_by_id(user.id).muted = result
```
Comment (BjoernPetersen) [inline dicers_bot/bot.py]:
```suggestion
            user.muted = result
```

File: dicers_bot/bot.py [hunk 6]
```diff
@@ -128,41 +171,58 @@ def remind_users(self, update: Update = None) -> bool:
         return all([bool(chat.show_attend_keyboard()) for chat in self.chats.values()])
 
     def handle_attend_callback(self, update: Update):
+        tz = timezone("Europe/Berlin")
+        latest_reply_time = tz.localize(datetime.today().replace(hour=20, minute=30))
```
Comment (BjoernPetersen) [inline dicers_bot/bot.py]: We shouldn't

File: dicers_bot/bot.py [hunk 7]
```diff
@@ -128,41 +171,58 @@ def remind_users(self, update: Update = None) -> bool:
...
         attends = callback.data == "attend_True"
+        if attends:
+            chat.current_event.add_attendee(user)
+            self.unmute_user(chat.id, user)
+        else:
+            chat.current_event.add_absentee(user)
+            try:
+                chat.current_event.remove_attendee(user)
+                until_date = timedelta(hours=1)
+                self.mute_user(chat.id, user, until_date)
```
Comment (BjoernPetersen) [inline dicers_bot/bot.py]: Maybe wait for a fixed amount of time so the user has a chance to give a reason (as previously discussed in #16)

File: dicers_bot/bot.py [hunk 8]
```diff
@@ -128,41 +171,58 @@ def remind_users(self, update: Update = None) -> bool:
...
-            else:
-                chat.current_event.add_absentee(user)
+            self.save_state()
+            chat.update_attend_message()
         except Exception as e:
-            print("Exception: {}".format(e))
-
-        self.save_state()
-        chat.update_attend_message()
+            print(e)
```
Comment (BjoernPetersen) [inline dicers_bot/bot.py]: Evil

File: dicers_bot/chat.py [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+
+    def _build_attend_message(self):
+        self.logger.info("Build attend message for event: {}".format(self.current_event))
+        message = "Wer ist dabei?"
+        attendees = self.current_event.attendees
+        absentees = self.current_event.absentees
+
+        if attendees:
+            self.logger.info("attend message has attendees")
+            message += " Bisher: " + ", ".join([user.name for user in attendees])
```
Comment (BjoernPetersen) [inline dicers_bot/chat.py]: Prepend a new line here

File: dicers_bot/user.py [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+    def __hash__(self):
+        return self.name.__hash__()
+
+    def __str__(self) -> str:
+        return "{} ({}{})".format(self.name, self.roll, "+1" if self.jumbo else "")
+
+    @classmethod
+    def from_tuser(cls, chat_user: TUser):
+        user = User(chat_user.name, chat_user.id, chat_user)
+
```
Comment (BjoernPetersen) [inline dicers_bot/user.py]: 😡 Remove newline

File: main.py [hunk 11]
```diff
@@ -34,33 +54,33 @@ def start(token: str):
...
+    )
 
     state_file = "state.json"
     if os.path.exists(state_file):
         with open(state_file) as f:
             try:
                 state = json.load(f)
             except json.decoder.JSONDecodeError as e:
-                print("Unable to load previous state: {}".format(e))
+                logging.Logger("state.json").warning(f"Unable to load previous state: {e}")
```
Comment (BjoernPetersen) [inline main.py]: Use `createLogger`

File: main.py (hunk 4 above)
Comment (Chabare) [reply main.py]: 795b36a2faf4be907be2efee98cabff38e6076d0

File: dicers_bot/bot.py (hunk 5 above)
Comment (Chabare) [reply dicers_bot/bot.py]: 5691658ec5e9fa016c72161654a8489cea022c03

File: dicers_bot/bot.py (hunk 6 above)
Comment (Chabare) [reply dicers_bot/bot.py]: 7e09b0c031b8783890bca04b0dc035abc5f79b3e

File: dicers_bot/bot.py (hunk 8 above)
Comment (Chabare) [reply dicers_bot/bot.py]: 7f006621aa6f785b93706d137f9fedd6c5d72c9f

File: dicers_bot/chat.py (hunk 9 above)
Comment (Chabare) [reply dicers_bot/chat.py]: 007a70d8d7abd61b794b283ccc997cd6eb6f69a9

File: dicers_bot/user.py (hunk 10 above)
Comment (Chabare) [reply dicers_bot/user.py]: Nope

File: main.py (hunk 11 above)
Comment (Chabare) [reply main.py]: 8e86a39055912b3be79f86487ca84faef26ccbf5

File: dicers_bot/bot.py [hunk 12]
```diff
@@ -97,26 +105,61 @@ def register_main(self, update: Update):
...
+            if not chat.get_user_by_id(update.message.from_user.id):
+                user = User.from_tuser(update.message.from_user)
+                chat.add_user(user)
+
+        return user
+
+    def restrict_user(self, chat_id: str, user: User, until_date: timedelta, **kwargs):
+        until_date = datetime.now() + until_date
+        try:
+            result = self.updater.bot.restrict_chat_member(chat_id, user.id, until_date=until_date, **kwargs)
```
[the remark this answers is not in the document]
Comment (Chabare) [reply dicers_bot/bot.py]: `until_date` actually accepts both an int and a datetime object. 

```until_date (:obj:`int` | :obj:`datetime.datetime`, optional): Date when restrictions
                will be lifted for the user, unix time. If user is restricted for more than 366
                days or less than 30 seconds from the current time, they are considered to be
                restricted forever.```

I changed the argument into a timestamp anyway ().

File: dicers_bot/bot.py (hunk 7 above)
Comment (Chabare) [reply dicers_bot/bot.py]: 78d7b801eda8eb4a43fc097ab1ca2fd02affa2d9

**Pull request merged**